libkatana_python_native/include/katana/python/TypeTraits.h

```diff
@@ -0,0 +1,79 @@
+#ifndef KATANA_LIBKATANAPYTHONNATIVE_KATANA_PYTHON_TYPETRAITS_H_
+#define KATANA_LIBKATANAPYTHONNATIVE_KATANA_PYTHON_TYPETRAITS_H_
+
+#include <pybind11/pybind11.h>
+
+namespace katana {
+
+template <typename T, typename Enable = void>
+struct PythonTypeTraits {};
+
+template <typename T>
+struct PythonTypeTraits<T, std::void_t<decltype(pybind11::type::of<T>())>> {
+  /// All pybind11 types use void* at the ctypes level.
+  static pybind11::object ctypes_type() {
+    pybind11::module_ ctypes = pybind11::module_::import("ctypes");
+    return ctypes.attr("c_void_p");
+  }
+
+  /// Get a Python object which represents the type in Python for purposes of
+  /// numba or type checking.
+  static pybind11::object representation() { return pybind11::type::of<T>(); }
+};
+
+#define PYTHON_TYPE_TRAITS(T, numpy_name, ctypes_name)                         \
+  template <>                                                                  \
```

Comment:
Why the empty template?

Reply:
It's how you write fully specialized template specializations in C++. Otherwise it would be a new template instead of a specialization.

Reply:
Forgot about good ol' template specialization.